drivers: counter: add set value and 64b ticks

XenuIsWatching opened the pull request:
This also introduces an api for set the counter time.

Also, this set frequency to be 64 bits rather than 32 bits, as counters,
such as those that use fractional adders, tick at resolutions
greater than 4294967295Hz. This changes the freq in the common info
to a uint64_t. This retains the existing counter_get_freq function which
will still return a 32bit value. This also adds a counter_get_freq_64
function to get a 64bit value for the frequency. Unfortunately, I do not
have a 'public' driver that uses this.

Add inline helper functions for converting ticks to nanoseconds.

Also implement an inline helper for converting microseconds or
nanoseconds with a 64bit tick value.

_this was a split from https://github.com/zephyrproject-rtos/zephyr/pull/89127_

XenuIsWatching committed:
drivers: counter: stm32: add counter reset

Add a reset counter api to set the time back to 0.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: counter: add kconfig for enabling 64b frequencies

Some counters, such as those that use fractional adders, tick at
resolutions greater than 4294967295Hz. This changes the freq in the common
info to a uint64_t. This adds a KConfig that can be selected by a driver.

This also has the counter ticks to sec (and vice versa) helper functions
call the get frequency for 64 bits if the kconfig is selected.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

mathieuchopstm commented on include/zephyr/drivers/counter.h:
```diff
@@ -196,7 +196,7 @@ struct counter_config_info {
 	 * Frequency of the source clock if synchronous events are counted.
 	 */
-	uint32_t freq;
+	uint64_t freq;
```
Nit: I would move this member at the start of the structure for alignment/packing reasons.

XenuIsWatching replied:
yeah, that makes sense... save 4 bytes of memory

XenuIsWatching replied:
done

etienne-lms reviewed (commented):
I think commit `drivers: counter: stm32: introduce counter set value api` should be split in 2 commits: one for the generic part (*counter.h*) and one for the STM32 driver. That is not a strong opinion, maybe maintainer is fine as-is.

XenuIsWatching commented:
@etienne-lms looking through your comments on formatting, i see the same issues ALL throughout this `counter.h`, I'll just add a commit and the head which "runs clang-format" on `counter.h`

XenuIsWatching committed:
drivers: counter: introduce counter set value api

Introduce a counter set value api to set the ticks for 32b and 64b.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: counter: stm32: implement counter set value api

Implement the counter_set_value api for the STM32.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

etienne-lms reviewed (dismissed):
For commit "drivers: counter: run clang-format on counter.h" I leave it to maintainer to tell if the commit is worth it.
LGTM.

XenuIsWatching commented:
something I just realized is that the `alarm` callbacks will need a 64b tick value (as well as likely the `top` values)

XenuIsWatching commented:
> > something I just realized is that the `alarm` callbacks will need a 64b tick value (as well as likely the `top` values)
> 
> thinking even more bout this... it gets rather complicated for counter drivers which do not use a 64b tick register and to support 64b across the board (top value, alarm ticks, values, max top values)... I'm beginning to think maybe it would be better to just have a KConfig (maybe `CONFIG_COUNTER_64BITS`) that selects a typedef for either a `uint32_t` or a `uint64_t` such as `counter_ticks_t` which is then used throughout the driver, rather than having to worry about if you need to use the 32b api or the 64b api and having 32b or 64b in some alarm callbacks (and others)
> 
> I also can't see a reason to really still have the 32b apis around if you know you've got a 64b counter
> 
> such as
> 
> ```c
> typedef int (*counter_api_set_value)(const struct device *dev, uint32_t ticks);
> ```
> 
> would just simply become
> 
> ```c
> typedef int (*counter_api_set_value)(const struct device *dev, counter_ticks_t ticks);
> ```
> 
> as well as
> 
> ```c
> typedef void (*counter_alarm_callback_t)(const struct device *dev, uint8_t chan_id, uint32_t ticks,
> 					 void *user_data);
> ```
> 
> would become
> 
> ```c
> typedef void (*counter_alarm_callback_t)(const struct device *dev, uint8_t chan_id, counter_ticks_t ticks,
> 					 void *user_data);
> ```
> 
> ....and the other places where `ticks`, `top` values are defined...

I experimented here for it: https://github.com/XenuIsWatching/zephyr/commit/7948d9d92b80303a4e6573ee7e9b3dc29cff41a0

@nordic-krch Would you think this would be a better approach?

XenuIsWatching commented:
> That sounds ok, there will be some additional code needed in 32 bit counters when there are mixture of 64 bit/ 32 bit counters in the system but that should not be big.

alrighty, I'll bring that commit in to this PR

XenuIsWatching commented on include/zephyr/drivers/counter.h:
```diff
@@ -300,7 +308,35 @@ static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
+	if (config->freq) {
+		return config->freq > UINT32_MAX ? UINT32_MAX : (uint32_t)config->freq;
+	} else {
+		return api->get_freq(dev);
```
done

nordic-krch reviewed (dismissed):
Look ok.

XenuIsWatching commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -309,20 +309,12 @@ tick_t rtc_stm32_read(const struct device *dev)
+static int rtc_stm32_get_value(const struct device *dev, counter_ticks_t *ticks)
 {
-	*ticks = (uint32_t)rtc_stm32_read(dev);
+	*ticks = (counter_ticks_t)rtc_stm32_read(dev);
```
done

etienne-lms commented on drivers/counter/Kconfig:
```diff
@@ -22,6 +22,12 @@ config COUNTER_SHELL
+config COUNTER_64BITS
+	bool
+	help
+	  This defines the counter ticks values (counter_ticks_t) to
```
For consistency with other hidden symbols/options descriptions:
```suggestion
	  Set by counter drivers to define counter ticks values (counter_ticks_t) to

etienne-lms commented on drivers/counter/counter_ite_it8xxx2.c:
```diff
@@ -283,7 +283,7 @@ static int counter_it8xxx2_init(const struct device *dev)
 	uint8_t et8_ctrl = counter_it8xxx2_read8(dev, ET8CTRL);
 
 	LOG_DBG("max top value = 0x%08x", config->info.max_top_value);
```
Cast to `(uint32_t)`?
If several counter drivers are enabled and 1 sets `CONFIG_COUNTER_64BITS=y`, then all these ticks related are 64bit in which case `%x` may taint compilers.

etienne-lms commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -739,9 +725,6 @@ static DEVICE_API(counter, rtc_stm32_driver_api) = {
 	.start = rtc_stm32_start,
 	.stop = rtc_stm32_stop,
 	.get_value = rtc_stm32_get_value,
-#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS
```
A few line above, i think `rtc_stm32_config::counter_info::max_top_value should be `UINT64_MAX` when `CONFIG_COUNTER_RTC_STM32_SUBSECONDS| is enabled:
```patch
 static const struct rtc_stm32_config rtc_config = {
 	.counter_info = {
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS
+		.max_top_value = UINT64_MAX,
+#else
 		.max_top_value = UINT32_MAX,
+#endif /* CONFIG_COUNTER_RTC_STM32_SUBSECONDS */

etienne-lms commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -237,12 +231,12 @@ static int rtc_stm32_stop(const struct device *dev)
 
 #if !defined(COUNTER_NO_DATE)
-tick_t rtc_stm32_read(const struct device *dev)
+counter_tick_t rtc_stm32_read(const struct device *dev)
```
Replace `counter_tick_t` with `counter_ticks_t` in this whole source file. Error reported by CI build tests.

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -35,6 +35,18 @@
+ * @brief Define the typedef to be 32b or 64b for ticks
+ *
+ * It is still okay for a driver or application to still use `uint32_t` or
+ * `uint64_t` if it is to always expected to be that.
```
It seems not. CI reports build failures on non matching handler prototypes/signatures:
```
/__w/zephyr/zephyr/drivers/counter/counter_esp32_tmr.c:320:22: error: initialization of 'int (*)(const struct device *, counter_ticks_t *)' {aka 'int (*)(const struct device *, long long unsigned int *)'} from incompatible pointer type 'int (*)(const struct device *, uint32_t *)' {aka 'int (*)(const struct device *, unsigned int *)'} [-Werror=incompatible-pointer-types]
  320 |         .get_value = counter_esp32_get_value,
      |                      ^~~~~~~~~~~~~~~~~~~~~~~

XenuIsWatching replied:
likely that was just an issue with the rebase I did

sylvioalves commented on drivers/counter/Kconfig:
```diff
@@ -22,6 +22,12 @@ config COUNTER_SHELL
+config COUNTER_64BITS
+	bool
+	help
+	  Set by counter drivers to defines the counter ticks
```
`to define` or
```
	  Select this option if the counter driver should use 64-bit
	  tick values for counter_ticks_t instead of 32-bit.
	  This is typically set by drivers that require extended
	  tick ranges beyond the 32-bit limit.
```

XenuIsWatching replied:
done

XenuIsWatching commented on include/zephyr/drivers/counter.h:
```diff
@@ -330,27 +344,85 @@ static inline uint32_t z_impl_counter_us_to_ticks(const struct device *dev,
+#ifndef CONFIG_COUNTER_64BITS
+static inline counter_ticks_t z_impl_counter_us_to_ticks(const struct device *dev, uint64_t us)
+{
+	uint64_t ticks = (us * z_impl_counter_get_frequency_64(dev)) / USEC_PER_SEC;
```
Frequency is different than ticks. It needs to get the frequency in 64bits so it doesn't overflow.

sylvioalves replied:
Yeah, I know. But with the changes, seems `z_impl_counter_get_frequency()` is not used anymore and could go. Is it?

XenuIsWatching commented:
@nordic-krch and @etienne-lms, Are you able to resolve your comments?

sylvioalves commented:
@XenuIsWatching, I get a few build warnings when running `west build -p -b esp32s3_devkitc/esp32s3/procpu samples/drivers/counter/alarm/`. Would you check that sample and provide necessary changes as well?

```c
zephyr/samples/drivers/counter/alarm/src/main.c:110:39: warning: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'counter_ticks_t' {aka 'long long unsigned int'} [-Wformat=]
  110 |         printk("Set alarm in %u sec (%u ticks)\n",
  ```

XenuIsWatching commented:
sorry @nordic-krch , could you take a look again, there was a minor issue where I had to swap the order a struct for better byte packing in `struct counter_alarm_cfg` with 64b ticks

kartben reviewed (dismissed):
I don't think it's ok to break API for a driver that, although apparently marked as "0.8.0 / Unstable", is de-facto "Stable" IMO (note the "`@since 1.14`").
At a _very minimum_, this requires some explanation in the migration guide but I would really treat this as a breaking change that will break drivers downstream.

XenuIsWatching commented:
> I don't think it's ok to break API for a driver that, although apparently marked as "0.8.0 / Unstable", is de-facto "Stable" IMO (note the "`@since 1.14`"). At a _very minimum_, this requires some explanation in the migration guide but I would really treat this as a breaking change that will break drivers downstream.

It shouldn't break someone who is still using the 32b... but it could cause some impact for those using the 64b. I can add this note to the migration guide
```
* Drivers that support 64-bit ticks now can use the new KConfig :kconfig:option:`CONFIG_COUNTER_64BIT` to enable
  this feature. The drivers that support this feature are :dtcompatible:`intel,ace-art-counter`,
  :dtcompatible:`intel,ace-rtc-counter`, :dtcompatible:`ti,cc23x0-rtc`, :dtcompatible:`espressif,esp32-counter`
  and the :dtcompatible:`st,stm32-rtc` if :kconfig:option:`CONFIG_COUNTER_RTC_STM32_SUBSECONDS` is enabled. This moves the `uint32_t` previously used for ticks to a typedef
  `counter_ticks_t` that is `uint64_t` when 64-bit ticks are enabled and `uint32_t` otherwise.
  The API functions that use ticks have been updated to use this typedef. This change is
  backward compatible as long as the application does not enable 64-bit ticks. Enabling
  64-bit ticks may require changes in the application code to use the new typedef and to
  handle the larger tick values. (:github:`92736`)
  ```
  @kartben How's this sound?

kartben commented:
> > I don't think it's ok to break API for a driver that, although apparently marked as "0.8.0 / Unstable", is de-facto "Stable" IMO (note the "`@since 1.14`"). At a _very minimum_, this requires some explanation in the migration guide but I would really treat this as a breaking change that will break drivers downstream.
> 
> It shouldn't break someone who is still using the 32b... but it could cause some impact for those using the 64b. I can add this note to the migration guide
> 
> ```
> COUNTER
> ===
> 
> * Drivers that support 64-bit ticks now can use the new KConfig
>   :kconfig:option:`CONFIG_COUNTER_64BIT` to enable this feature. The drivers that support
>   this feature are :dtcompatible:`intel,ace-art-counter`,
>   :dtcompatible:`intel,ace-rtc-counter`, :dtcompatible:`ti,cc23x0-rtc`,
>   :dtcompatible:`espressif,esp32-counter`, and the :dtcompatible:`st,stm32-rtc` if
>   :kconfig:option:`CONFIG_COUNTER_RTC_STM32_SUBSECONDS` is enabled. This moves the `uint32_t`
>   previously used for ticks to a typedef `counter_ticks_t` that is `uint64_t` when 64-bit
>   ticks are enabled and `uint32_t` otherwise. The API functions that use ticks have been
>   updated to use this typedef. This change is backward compatible as long as the application
>   does not enable 64-bit ticks. Enabling 64-bit ticks may require changes in the application
>   code to use the new typedef and to handle the larger tick values. (:github:`94189`)
> ```
> 
> @kartben How's this sound?

my main concern is folks _implementing_ drivers downstream -- you've basically removed the `get_value_64` field in the API struct so it will fail hard on them, so that's I think where the migration guide entry would need to focus.. But again, this has been there for 3+ years so it doesn't feel right to just drop it. My impression is that Counter subsystem is operating as if it were a stable API (and we should probably make it official if that's the case) hence why I am bringing this up.

jhedberg commented:
@XenuIsWatching there are some merge conflicts which require a rebase

kartben commented:
> [...] it just seems odd and rather useless in first place as having something 64b ticks supported and then just returning 32b ticks within alarm and top values when they really can be 64b really seems odd.

I don't necessarily disagree that the API initially introduced was probably poorly designed, but that doesn't make the change "less" breaking :/ (just similar to if we rename an API that has a typo in it... this does fix the oddity, but breaks folks).

XenuIsWatching commented:
> @XenuIsWatching there are some merge conflicts which require a rebase



> > [...] it just seems odd and rather useless in first place as having something 64b ticks supported and then just returning 32b ticks within alarm and top values when they really can be 64b really seems odd.
> 
> I don't necessarily disagree that the API initially introduced was probably poorly designed, but that doesn't make the change "less" breaking :/ (just similar to if we rename an API that has a typo in it... this does fix the oddity, but breaks folks).

I understand that, but it should never be too late to do it right :)

XenuIsWatching commented:
> @XenuIsWatching there are some merge conflicts which require a rebase

done

sylvioalves commented:
> @kartben @sylvioalves @XenuIsWatching I am a bit confused. As long as users do _not_ enable `CONFIG_COUNTER_64BITS`, is this really a breaking change?

1) It is setting `select COUNTER_64BITS` to all drivers that supports 64-bits, no user inputs.
2) For the same drivers, it is removing the `.get_value_64` API and defaulting to `.get_value`.

It means that any downstream apps already using `.get_value_64` will fail without migrating properly. 
See https://github.com/zephyrproject-rtos/zephyr/pull/94189#issuecomment-3408444669

kartben commented:
> @kartben based on your comments, should this be tagged with "breaking API changes" and also added into dev review or TSC meeting? Asking this due to upcoming feature freeze.

Yeah I think it needs to be discussed in the architecture WG meeting. We recently had a similar situation where someone wanted introduced a breaking change to zbus which, at the time, was also still experimental-on-paper-but-de-facto-stable-for-users, and that breaking API change was reverted + zbus effectively graduated to stable. https://github.com/zephyrproject-rtos/zephyr/pull/88834
I think we might be looking at doing something similar here, i.e. Counter API should really be considered 1.0 IMO.

XenuIsWatching committed:
drivers: counter: add 64bit apis

Add apis for 64 bits.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

nordic-krch commented on include/zephyr/drivers/counter.h:
```diff
@@ -210,35 +223,106 @@ struct counter_config_info {
+
+/** @brief Alarm callback structure.
+ */
+struct counter_alarm_cfg_64 {
```
would it be possible to keep a single struct backward compatible with anonymous unions? Same for `struct counter_top_cfg`

```
struct counter_alarm_cfg {
    union {
       uint32_t ticks;
#ifdef CONFIG_COUNTER_64BITS
       uint64_t ticks_64;
#endif
     }
    union {
       counter_alarm_callback_t callback;
#ifdef CONFIG_COUNTER_64BITS
       counter_alarm_callback_64_t callback_64;
#endif
     }
     void *user_data;
     uint32_t flags;
}
```

nordic-krch reviewed (commented):
I tried to compile with and without changes for nrf54l15dk. I would expect no code size increase but I see some:
- API structure +32 bytes due to lack of ifdef
- config structure +8 bytes per instance
- `z_impl_counter_get_frequency` got changed to `z_impl_counter_get_frequency_64` and it is +20 bytes per each call.

Can you try to make it so there is no change in 32 bit implementations?

XenuIsWatching committed:
drivers: counter: add ns to ticks helper

There is already a helper for converting ticks to ns. Add one
that does the opposite.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: counter: add inline helpers for(n/u)sec to ticks in 64b

Add inline helpers to convert nanoseconds or microseconds to 64b ticks
and vice versa.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: counter: run clang-format on counter.h

Run clang-format on counter.h

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented on include/zephyr/drivers/counter.h:
```diff
@@ -332,10 +497,9 @@ static inline uint32_t z_impl_counter_us_to_ticks(const struct device *dev,
+static inline uint64_t z_impl_counter_ticks_to_us(const struct device *dev, uint32_t ticks)
 {
-	return ((uint64_t)ticks * USEC_PER_SEC) / z_impl_counter_get_frequency(dev);
+	return ((uint64_t)ticks * USEC_PER_SEC) / z_impl_counter_get_frequency_64(dev);
```
> yeah, maybe 64b freq should be a Kconfig. Zephyr after all mainly targets embedded and >4GHz frequencies are not that common.

@nordic-krch alright, I made it kconfig take a look 😄

kartben commented on include/zephyr/drivers/counter.h:
```diff
@@ -210,35 +227,106 @@ struct counter_config_info {
 	counter_api_start start;
 	counter_api_stop stop;
 	counter_api_get_value get_value;
-	counter_api_get_value_64 get_value_64;
```
This field now being gated by a Kconfig means we still have a breaking change for downstream drivers (as illustrated by the update required for in-tree drivers)

XenuIsWatching replied:
then it will just be added to the migration guide

kartben replied:
> yeah... we already had that discussion in the arch meeting. This API is not technically stable, and changes can still happen.
> 
> I added a note to the migration guide for 4.4

I believe the conclusion of the discussion was that we actually wanted to treat it as stable, and as a matter of fact it now is https://github.com/zephyrproject-rtos/zephyr/pull/98919

XenuIsWatching committed:
doc: migration-guide-4.4: counters that implement get_value_64

Add a migration note for counter drivers that use `get_value_64` that
they need to select CONFIG_COUNTER_64BITS_TICKS.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

henrikbrixandersen commented:
Architecture WG 2025-12-09:
- @XenuIsWatching presented the proposed API change (guarding the existing `counter_api_get_value_64 get_value_64` behind the Kconfig option `CONFIG_COUNTER_64BITS_TICKS` being introduced with https://github.com/zephyrproject-rtos/zephyr/pull/94189).
- No objections from the Architecture WG.

XenuIsWatching commented:
@nordic-krch @etienne-lms Please review 😄

XenuIsWatching commented:
@nordic-krch sorry, there was a build issue. I had to repush. Please approve again. Thanks!

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -212,13 +227,76 @@ struct counter_config_info {
+	 * It can be relative (to now) or an absolute value (see @ref
+	 * COUNTER_ALARM_CFG_ABSOLUTE). Both, relative and absolute, alarm
+	 * values can be any value between zero and the current top value (see
+	 * @ref counter_get_top_value). When setting an absolute alarm value
```
```suggestion
	 * @ref counter_get_top_value_64). When setting an absolute alarm value

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -698,18 +915,297 @@ static inline int z_impl_counter_set_guard_period(const struct device *dev,
+ * @param cfg		Configuration. Cannot be NULL.
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP if request is not supported (e.g. top value cannot be
```
I think this should be `-EINVAL` (in which case `counter_set_top_value_64()` inline description should be also changed).

henrikbrixandersen commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -99,6 +99,10 @@ Controller Area Network (CAN)
+* Drivers that implement ``get_value_64`` API will now need to select
+  :kconfig:option:`CONFIG_COUNTER_64BITS_TICKS` to enable the API.
+  (:github:`94189`).
+
```
Shouldn't the application select this? Not the driver? Not sure I follow the logic here.

What if an application wants to use the 32-bit counter API with a counter driver supporting 64 bit wide counters (e.g. for backwards or cross-SoC compatibility)?

I thought the idea was to implement 32-bit API wrappers for the 64-bit capable counter drivers?

henrikbrixandersen replied:
> > Shouldn't the application select this? Not the driver? Not sure I follow the logic here.
> 
> The driver should select this as it understands the physical reality of the hardware wither that is the RP2350, ESP32, or whatever, this was also discussed in the review earlier as it is kconfig that isn't and shouldn't really selectable from an application. [#94189 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/94189#discussion_r2395443770)

I wouldn't call a single comment a discussion. I don't see a reason for including the overhead of the 64-bit logic, API calls, etc. unless an application wishes to make use of it. Making this optional and selectable from the application aligns with how API extensions are handled in other driver APIs (see e.g. how `CONFIG_CAN_FD_MODE` is done).

It was assumption/impression from our discussion at the Architecture WG meeting, that this API extension would be optional.

> > What if an application wants to use the 32-bit counter API with a counter driver supporting 64 bit wide counters (e.g. for backwards or cross-SoC compatibility)?
> 
> The existing 32-bit API functions (like `counter_get_value()`) remain available. If a driver is "64-bit capable," it still implements the standard 32-bit counter API. The 64-bit additions are extensions (e.g., `counter_get_value_64()`). An application can continue to call the 32-bit functions on a 64-bit driver without any changes. The driver simply truncates or handles the 32-bit request as it did before. See how this is done on the `counter_esp32` in this PR. It's just if an already exiting driver that implements the 64-bit (very limited existing and rather weak) implementation, then will need to add the `select COUNTER_64BITS_TICKS` to their driver.

But it still adds overhead if the application is not using the 64-bit API extension.

XenuIsWatching replied:
> It was assumption/impression from our discussion at the Architecture WG meeting, that this API extension would be optional.

> But it still adds overhead if the application is not using the 64-bit API extension.

If an SoC has a 64-bit counter (like the RP2350 or ESP32), the driver should use 64-bit types to interact with it natively. Forcing the driver to treat 64-bit hardware as 32-bit is just not a real world thing. While I understand you point of trying to just 'save a kilobyte', I just really don't see this as something worth while for a non-existant use case. Also, there are existing drivers such as the `counter_ace_v1x_art` (and others) will ONLY work with this select. The optionally of this is that this gives NO overhead (and actually rather remove some overhead) for existing 32b SoC IPs. All the 32b APIs will work as is even if the application is already written to use that 32b api.

I'm going to leave to @etienne-lms and @nordic-krch if they feel this is a reasonable thing as they are the mainters of the counter_esp32 and the counter subsystem here. Furthermore, I am not saying this is not possible, as it would be possible to change just the `counter_esp32` (as it is only in-tree one that has both 32b and 64b api) `selects` to an `imply` (as if the hardware is 64b as definied by the SoC, it mostly certainly should be **expected** to return 64b) and just a conditional compile those in to the drivers (some would MUST remain as `select` as they do not implement the 32b api and I am NOT going to go and change those drivers as they have already always been implemented)... but thats an effort that I do not see having a real world use case.
The reason why i'm rather pushing back here on doing 'work' here in the short-term as I am on on paternity leave and rather difficult for me to 'come out' of the leave

sylvioalves replied:
I think @henrikbrixandersen is correct.

COUNTER_64BITS_TICKS should be application-selectable, not driver-selected. It controls API overhead, not only hardware capability. Applications that don't need 64-bit tick ranges shouldn't pay for the extra function pointers and code. Drivers with 64-bit hardware can still work correctly with the 32-bit API by truncating values internally.
Another example is `CONFIG_FPU`, which is not enabled by default and yet is a HW feature.

> If an SoC exposes a native 64-bit counter (e.g. RP2350, ESP32), the driver should interact with it using 64-bit types. Forcing such hardware through a 32-bit abstraction does not reflect real-world usage and provides no practical benefit.

That is true. It is 64-bit but it casts to uint32 to meet the 32-bit API. Even so I still think the suggested approach is the right one.

@nordic-krch, what your thoughts?

XenuIsWatching committed:
drivers: counter: add kconfig for enabling 64b tick api

Add a kconfig to be used as a flag for indicating if it supports 64b
ticks and another flag which will enable the 64b api. The kconfig
indicating that it supports 64b is a dependency of the kconfig that
will enable the api.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>
